pkg/broker/handler/processors/deliver/processor.go

```diff
@@ -101,41 +104,69 @@ func (p *Processor) Process(ctx context.Context, event *event.Event) error {
 	}
 
 	// Forward the event copy that has hops removed.
-	resp, res := p.DeliverClient.Request(cecontext.WithTarget(dctx, target.Address), copy)
-	if !protocol.IsACK(res) {
+	if err := p.deliver(dctx, target, broker, (*binding.EventMessage)(&copy), hops); err != nil {
 		if !p.RetryOnFailure {
-			return fmt.Errorf("target delivery failed: %v", res.Error())
+			return err
 		}
 
-		logging.FromContext(ctx).Warn("target delivery failed", zap.String("target", tk), zap.String("error", res.Error()))
+		logging.FromContext(ctx).Warn("target delivery failed", zap.String("target", tk), zap.Error(err))
 		return p.sendToRetryTopic(ctx, target, event)
```

Comment:
This is a comment on the [above strategy](https://github.com/google/knative-gcp/pull/1135/files#diff-b746f19389275f3eb9488200919af700R93) of cloning the event to remove the hops attribute and keeping the original around for possible retry. I'm curious why this cloning is necessary. Could we remove the hops attribute from `event` and add it back if needed? Is there an issue with concurrent access?

I know this isn't really part of this PR so feel free to ignore.

Reply:
Yes, this event may be fanned out to multiple processors and it would not be safe to mutate the event here. My next idea for optimization was to write a message transformer to mutate hops when writing the outgoing http message rather than cloning the event. I will start work on that optimization once this PR is merged.